File: scripts/item-generator/assets/material-paths.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Ian Lucas. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { basename } from "path";
import { ensure } from "../../../src/utils.ts";
import { type Cs2Runtime } from "../../cs2-tools/types.ts";

export type MaterialReferenceResolver = (path: string) => string | undefined;

export function normalizeMaterialResourcePath(path: string): string {
    return path
        .replace(/\\/g, "/")
        .replace(/^resource(?:_name)?:/, "")
        .replace(/\/+/g, "/");
}

export function getPaintCompositeMaterialPath(className: string, compositeMaterialPath?: string): string {
    return compositeMaterialPath ?? `weapons/paints/legacy/${className}.vcompmat`;
}

export function toCompiledMaterialResourcePath(path: string): string {
    const normalized = normalizeMaterialResourcePath(path).toLowerCase();
    if (normalized.endsWith("_c")) {
        return normalized;
    }
    if (normalized.endsWith(".vcompmat") || normalized.endsWith(".vmat") || normalized.endsWith(".vtex")) {
        return `${normalized}_c`;
    }
    return normalized;
}

function toSourceMaterialResourcePath(path: string) {
    return path.endsWith("_c") ? path.slice(0, -2) : path;
}

export function resolveMaterialResourcePath(runtime: Cs2Runtime, path: string): string {
    const compiledPath = toCompiledMaterialResourcePath(path);
    if (runtime.vpkIndex.has(compiledPath)) {
        return toSourceMaterialResourcePath(compiledPath);
    }
    const name = basename(compiledPath);
    const matches = [...runtime.vpkIndex.keys()].filter((candidate) => basename(candidate) === name);
    if (matches.length === 1) {
        return toSourceMaterialResourcePath(matches[0]!);
    }
    if (matches.length > 1) {
        throw new Error(`Ambiguous VPK entry for '${compiledPath}': ${matches.join(", ")}`);
    }
    throw new Error(`VPK entry not found: ${compiledPath}`);
}

export function getCompositeMaterialFilename(vcompmatPath: string, crc: string): string {
    return `${basename(normalizeMaterialResourcePath(vcompmatPath), ".vcompmat")}_${crc}.vcompmat.json`;
}

export function getVmatFilename(vmatPath: string, crc: string): string {
    return `${basename(normalizeMaterialResourcePath(vmatPath), ".vmat")}_${crc}.vmat.json`;
}

export function getTextureFilename(vtexPath: string, crc: string, extension: ".webp" | ".exr"): string {
    return `${basename(normalizeMaterialResourcePath(vtexPath), ".vtex")}_${crc}${extension}`;
}

export function getIndexedCompositeMaterialFilename(runtime: Cs2Runtime, vcompmatPath: string): string {
    const resolvedPath = resolveMaterialResourcePath(runtime, vcompmatPath);
    const vpkPath = toCompiledMaterialResourcePath(resolvedPath);
    const entry = ensure(runtime.vpkIndex.get(vpkPath), `VPK entry not found: ${vpkPath}`);
    return getCompositeMaterialFilename(resolvedPath, entry.crc);
}

export function getIndexedVmatFilename(runtime: Cs2Runtime, vmatPath: string): string {
    const resolvedPath = resolveMaterialResourcePath(runtime, vmatPath);
    const vpkPath = toCompiledMaterialResourcePath(resolvedPath);
    const entry = ensure(runtime.vpkIndex.get(vpkPath), `VPK entry not found: ${vpkPath}`);
    return getVmatFilename(resolvedPath, entry.crc);
}

export function patchMaterialResourceReferences(
    value: unknown,
    resolveCompositeMaterial: MaterialReferenceResolver,
    resolveVmat: MaterialReferenceResolver,
    resolveTexture: MaterialReferenceResolver
): unknown {
    if (typeof value === "string") {
        const normalized = normalizeMaterialResourcePath(value);
        if (normalized.endsWith(".vcompmat")) {
            return ensure(resolveCompositeMaterial(normalized), `Unable to rewrite composite material reference: ${value}`);
        }
        if (normalized.endsWith(".vmat")) {
            return ensure(resolveVmat(normalized), `Unable to rewrite material reference: ${value}`);
        }
        if (normalized.endsWith(".vtex")) {
            return ensure(resolveTexture(normalized), `Unable to rewrite texture reference: ${value}`);
        }
        return value;
    }
    if (Array.isArray(value)) {
        return value.map((entry) =>
            patchMaterialResourceReferences(entry, resolveCompositeMaterial, resolveVmat, resolveTexture)
        );
    }
    if (value !== null && typeof value === "object") {
        return Object.fromEntries(
            Object.entries(value).map(([key, child]) => [
                key,
                patchMaterialResourceReferences(child, resolveCompositeMaterial, resolveVmat, resolveTexture)
            ])
        );
    }
    return value;
}

File: scripts/item-generator-cs/Assets/MaterialPaths.cs
using System.Text.RegularExpressions;

namespace ItemGenerator;

public static class MaterialPaths
{
    public static string NormalizeMaterialResourcePath(string path)
    {
        path = path.Replace('\\', '/');
        path = Regex.Replace(path, @"^resource(?:_name)?:", "");
        path = Regex.Replace(path, @"/+", "/");
        return path;
    }

    public static string GetPaintCompositeMaterialPath(string className, string? compositeMaterialPath)
    {
        return compositeMaterialPath ?? $"weapons/paints/legacy/{className}.vcompmat";
    }

    public static string GetStickerMaterialPath(string stickerMaterial)
    {
        return $"stickers/{stickerMaterial}.vmat";
    }

    public static string ToCompiledMaterialResourcePath(string path)
    {
        var normalized = NormalizeMaterialResourcePath(path).ToLowerInvariant();
        if (normalized.EndsWith("_c")) return normalized;
        if (normalized.EndsWith(".vcompmat") || normalized.EndsWith(".vmat") || normalized.EndsWith(".vtex"))
            return $"{normalized}_c";
        return normalized;
    }

    private static string ToSourceMaterialResourcePath(string path)
    {
        return path.EndsWith("_c") ? path[..^2] : path;
    }

    public static string ResolveMaterialResourcePath(ItemGeneratorContext ctx, string path)
    {
        var compiledPath = ToCompiledMaterialResourcePath(path);
        if (ctx.VpkIndex.ContainsKey(compiledPath))
            return ToSourceMaterialResourcePath(compiledPath);

        var name = Path.GetFileName(compiledPath);
        var matches = ctx.VpkIndex.Keys
            .Where(candidate => Path.GetFileName(candidate) == name)
            .ToList();

        if (matches.Count == 1)
            return ToSourceMaterialResourcePath(matches[0]);
        if (matches.Count > 1)
            throw new InvalidOperationException($"Ambiguous VPK entry for '{compiledPath}': {string.Join(", ", matches)}");

        throw new FileNotFoundException($"VPK entry not found: {compiledPath}");
    }

    public static string GetCompositeMaterialFilename(string vcompmatPath, string crc)
    {
        var baseName = Path.GetFileNameWithoutExtension(NormalizeMaterialResourcePath(vcompmatPath));
        if (baseName.EndsWith(".vcompmat")) baseName = baseName[..^9];
        return $"{baseName}_{crc}.vcompmat.json";
    }

    public static string GetVmatFilename(string vmatPath, string crc)
    {
        var baseName = Path.GetFileNameWithoutExtension(NormalizeMaterialResourcePath(vmatPath));
        if (baseName.EndsWith(".vmat")) baseName = baseName[..^5];
        return $"{baseName}_{crc}.vmat.json";
    }

    public static string GetTextureFilename(string vtexPath, string crc, string extension)
    {
        var baseName = Path.GetFileNameWithoutExtension(NormalizeMaterialResourcePath(vtexPath));
        if (baseName.EndsWith(".vtex")) baseName = baseName[..^5];
        return $"{baseName}_{crc}{extension}";
    }

    public static string GetIndexedCompositeMaterialFilename(ItemGeneratorContext ctx, string vcompmatPath)
    {
        var resolvedPath = ResolveMaterialResourcePath(ctx, vcompmatPath);
        var vpkPath = ToCompiledMaterialResourcePath(resolvedPath);
        if (!ctx.VpkIndex.TryGetValue(vpkPath, out var entry))
            throw new FileNotFoundException($"VPK entry not found: {vpkPath}");
        return GetCompositeMaterialFilename(resolvedPath, entry.Crc);
    }

    public static string GetIndexedVmatFilename(ItemGeneratorContext ctx, string vmatPath)
    {
        var resolvedPath = ResolveMaterialResourcePath(ctx, vmatPath);
        var vpkPath = ToCompiledMaterialResourcePath(resolvedPath);
        if (!ctx.VpkIndex.TryGetValue(vpkPath, out var entry))
            throw new FileNotFoundException($"VPK entry not found: {vpkPath}");
        return GetVmatFilename(resolvedPath, entry.Crc);
    }

    public static object? PatchMaterialResourceReferences(
        object? value,
        Func<string, string?> resolveCompositeMaterial,
        Func<string, string?> resolveVmat,
        Func<string, string?> resolveTexture)
    {
        if (value is string str)
        {
            var normalized = NormalizeMaterialResourcePath(str);
            if (normalized.EndsWith(".vcompmat"))
                return resolveCompositeMaterial(normalized)
                    ?? throw new InvalidOperationException($"Unable to rewrite composite material reference: {str}");
            if (normalized.EndsWith(".vmat"))
                return resolveVmat(normalized)
                    ?? throw new InvalidOperationException($"Unable to rewrite material reference: {str}");
            if (normalized.EndsWith(".vtex"))
                return resolveTexture(normalized)
                    ?? throw new InvalidOperationException($"Unable to rewrite texture reference: {str}");
            return str;
        }

        if (value is List<object?> list)
            return list.Select(entry => PatchMaterialResourceReferences(entry, resolveCompositeMaterial, resolveVmat, resolveTexture)).ToList();

        if (value is Dictionary<string, object?> dict)
            return dict.ToDictionary(kv => kv.Key,
                kv => PatchMaterialResourceReferences(kv.Value, resolveCompositeMaterial, resolveVmat, resolveTexture));

        return value;
    }
}
